funding/batch.go

```diff
@@ -0,0 +1,512 @@
+// channels within a single on-chain transaction.
+type Batcher struct {
+	cfg *BatchConfig
+
+	channels    []*batchChannel
+	lockedUTXOs []*walletrpc.UtxoLease
+
+	didPublish bool
+
+	wg           sync.WaitGroup
+	negotiateErr error
+}
+
+// NewBatcher returns a new batch channel funding helper.
+func NewBatcher(cfg *BatchConfig) *Batcher {
+	return &Batcher{
+		cfg: cfg,
+	}
+}
+
+// BatchFund starts the atomic batch channel funding process.
+//
+// NOTE: This method should only be called once per instance.
+func (b *Batcher) BatchFund(ctx context.Context,
+	req *lnrpc.BatchOpenChannelRequest) ([]*lnrpc.PendingUpdate, error) {
+
+	label, err := labels.ValidateAPI(req.Label)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse and validate each individual channel.
+	b.channels = make([]*batchChannel, 0, len(req.Channels))
+	for idx, rpcChannel := range req.Channels {
+		var pendingChanID [32]byte
+		if _, err := rand.Read(pendingChanID[:]); err != nil {
+			return nil, fmt.Errorf("error making temp chan ID: %v",
+				err)
+		}
+
+		fundingReq, err := b.cfg.RequestParser(&lnrpc.OpenChannelRequest{
+			SatPerVbyte:        uint64(req.SatPerVbyte),
+			NodePubkey:         rpcChannel.NodePubkey,
+			LocalFundingAmount: rpcChannel.LocalFundingAmount,
+			PushSat:            rpcChannel.PushSat,
+			TargetConf:         req.TargetConf,
+			Private:            rpcChannel.Private,
+			MinHtlcMsat:        rpcChannel.MinHtlcMsat,
+			RemoteCsvDelay:     rpcChannel.RemoteCsvDelay,
+			MinConfs:           req.MinConfs,
+			SpendUnconfirmed:   req.SpendUnconfirmed,
+			CloseAddress:       rpcChannel.CloseAddress,
+			FundingShim: &lnrpc.FundingShim{
+				Shim: &lnrpc.FundingShim_PsbtShim{
+					PsbtShim: &lnrpc.PsbtShim{
+						PendingChanId: pendingChanID[:],
+						NoPublish:     true,
+					},
+				},
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error parsing channel %d: %v",
+				idx, err)
+		}
+
+		// Prepare the stuff that we'll need for the internal PSBT
+		// funding.
+		fundingReq.PendingChanID = pendingChanID
+		fundingReq.ChanFunder = chanfunding.NewPsbtAssembler(
+			btcutil.Amount(rpcChannel.LocalFundingAmount), nil,
+			b.cfg.NetParams, false,
+		)
+
+		b.channels = append(b.channels, &batchChannel{
+			pendingChanID: pendingChanID,
+			fundingReq:    fundingReq,
+		})
+	}
+
+	// From this point on we can fail for any of the channels and for any
+	// number of reasons. This deferred function makes sure that the full
+	// operation is actually atomic: We either succeed and publish a
+	// transaction for the full batch or we clean up everything.
+	defer b.cleanup(ctx)
+
+	// Now that we know the user input is sane, we need to kick off the
+	// channel funding negotiation with the peers. Because we specified a
+	// PSBT assembler, we'll get a special response in the channel once the
+	// funding output script is known (which we need to craft the TX).
+	b.wg.Add(len(b.channels))
+	b.negotiateErr = nil
```

Comment:
An `errgroup` here could alternatively be used assuming we don't need to read `negotiateErr` in some other context. Also if we go this route, we won't need to worry about any synchronization between the goroutines to set the negotiation error.

Reply:
Ah, great suggestion! Not sure why I didn't do this in the first place.